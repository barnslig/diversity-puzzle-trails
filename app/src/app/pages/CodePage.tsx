import {
  AppBar,
  Box,
  Button,
  Container,
  IconButton,
  makeStyles,
  Toolbar,
  Typography,
} from "@material-ui/core";
import { Close } from "@material-ui/icons";
import { FormattedMessage, useIntl } from "react-intl";
import { Link } from "wouter";
import * as React from "react";

import ButtonProgressIndicator from "../../common/components/ButtonProgressIndicator";
import Code from "../../features/code/Code";
import StickyActionButtons from "../../common/components/StickyActionButtons";
import useCode from "../../common/hooks/api/useCode";
import useHandleApiError from "../../common/hooks/api/useHandleApiError";
import useSubmitCode from "../../common/hooks/api/useSubmitCode";

const useStyles = makeStyles((theme) => ({
  appBarTitle: {
    flexGrow: 1,
  },
}));

type CodePageProps = {
  codeId: string;
};

const CodePage = ({ codeId }: CodePageProps) => {
  const { data, error } = useCode(codeId);
  const classes = useStyles();
  const intl = useIntl();
  const [isLoading, submitCode] = useSubmitCode(codeId);

  const code = data?.data;

  const handleError = useHandleApiError();
  React.useEffect(() => handleError(error), [error]);

  const onSubmit = () => {
    if (!code) {
      return;
    }
    return submitCode(code);
  };

  return (
    <div>
      <AppBar position="fixed" color="inherit">
        <Toolbar>
          <Box marginRight={2}>
            <Link href="/">
              <IconButton
                edge="start"
                title={intl.formatMessage({
                  defaultMessage: "QR-Code nicht ausführen",
                  description:
                    "app bar close button of the confirm qr code action page",
                })}
              >
                <Close />
              </IconButton>
            </Link>
          </Box>
          <Typography
            className={classes.appBarTitle}
            component="h1"
            variant="h6"
          >
            <FormattedMessage
              defaultMessage="QR-Code Aktionen"
              description="title of the confirm qr code actions page"
            />
          </Typography>
        </Toolbar>
      </AppBar>
      <Box component="main" paddingTop={10}>
        <Container maxWidth="sm">
          <Typography gutterBottom variant="body1">
            <FormattedMessage
              defaultMessage="Dieser QR-Code führt folgende Aktionen aus:"
              description="body of the card to confirm a qr-code"
            />
          </Typography>

          <Code code={code} />

          <StickyActionButtons>
            <Button
              color="primary"
              fullWidth
              size="large"
              type="submit"
              variant="contained"
              disabled={!code || isLoading}
              onClick={onSubmit}
            >
              <FormattedMessage
                defaultMessage="Ausführen"
                description="confirm qr code button"
              />
              {isLoading && <ButtonProgressIndicator />}
            </Button>
          </StickyActionButtons>
        </Container>
      </Box>
    </div>
  );
};

export default CodePage;
