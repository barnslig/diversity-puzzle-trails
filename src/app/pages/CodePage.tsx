import {
  AppBar,
  Box,
  Button,
  CircularProgress,
  Container,
  IconButton,
  makeStyles,
  Toolbar,
  Typography,
} from "@material-ui/core";
import { Close } from "@material-ui/icons";
import { FormattedMessage, useIntl } from "react-intl";
import { Link, useLocation } from "wouter";
import { useSnackbar } from "notistack";
import * as React from "react";

import ApiError from "../../common/hooks/api/helper/ApiError";
import Code from "../../features/code/Code";
import config from "../../config";
import errorAwareFetcher from "../../common/hooks/api/helper/errorAwareFetcher";
import StickyActionButtons from "../../common/components/StickyActionButtons";
import useApiUrl from "../../common/hooks/api/useApiUrl";
import useCode from "../../common/hooks/api/useCode";
import useInstanceId from "../../common/hooks/useInstanceId";

const useStyles = makeStyles((theme) => ({
  appBarTitle: {
    flexGrow: 1,
  },
  buttonProgress: {
    position: "absolute",
    top: "50%",
    left: "50%",
    marginTop: -12,
    marginLeft: -12,
  },
}));

type CodePageProps = {
  codeId: string;
};

const CodePage = ({ codeId }: CodePageProps) => {
  const [, setLocation] = useLocation();
  const [isLoading, setIsLoading] = React.useState(false);
  const { enqueueSnackbar } = useSnackbar();
  const classes = useStyles();
  const { data, error } = useCode(codeId);
  const instanceId = useInstanceId();
  const intl = useIntl();
  const url = useApiUrl((gameId) => config.apiEndpoints.code(codeId, gameId));

  const code = data?.data;

  const handleError = (error?: ApiError) => {
    if (!error) {
      return;
    }

    const title = error.info?.errors[0].title;

    enqueueSnackbar(
      title
        ? title
        : intl.formatMessage({
            defaultMessage: "Ein Fehler ist aufgetreten!",
            description: "unknown error snack",
          }),
      { variant: "error" }
    );

    setLocation("/");
  };

  React.useEffect(() => handleError(error), [error]);

  const onSubmit = async () => {
    if (!url) {
      return;
    }

    setIsLoading(true);

    try {
      await errorAwareFetcher(() =>
        fetch(url, {
          method: "POST",
          headers: {
            Authorization: `Bearer ${instanceId}`,
          },
        })
      );
    } catch (error) {
      return handleError(error);
    }

    enqueueSnackbar(
      intl.formatMessage({
        defaultMessage: "QR-Code erfolgreich ausgef端hrt!",
        description: "success snack on exec qr code",
      }),
      { variant: "success" }
    );

    setLocation("/");
  };

  return (
    <div>
      <AppBar position="fixed" color="inherit">
        <Toolbar>
          <Box marginRight={2}>
            <Link href="/">
              <IconButton
                edge="start"
                title={intl.formatMessage({
                  defaultMessage: "QR-Code nicht ausf端hren",
                  description:
                    "app bar close button of the confirm qr code action page",
                })}
              >
                <Close />
              </IconButton>
            </Link>
          </Box>
          <Typography
            className={classes.appBarTitle}
            component="h1"
            variant="h6"
          >
            <FormattedMessage
              defaultMessage="QR-Code Aktionen"
              description="title of the confirm qr code actions page"
            />
          </Typography>
        </Toolbar>
      </AppBar>
      <Box component="main" paddingTop={10}>
        <Container maxWidth="sm">
          <Typography gutterBottom variant="body1">
            <FormattedMessage
              defaultMessage="Dieser QR-Code f端hrt folgende Aktionen aus:"
              description="body of the card to confirm a qr-code"
            />
          </Typography>

          <Code code={code} />

          <StickyActionButtons>
            <Button
              color="primary"
              fullWidth
              size="large"
              type="submit"
              variant="contained"
              disabled={!code || isLoading}
              onClick={onSubmit}
            >
              <FormattedMessage
                defaultMessage="Ausf端hren"
                description="confirm qr code button"
              />
              {isLoading && (
                <CircularProgress
                  size={24}
                  className={classes.buttonProgress}
                />
              )}
            </Button>
          </StickyActionButtons>
        </Container>
      </Box>
    </div>
  );
};

export default CodePage;
