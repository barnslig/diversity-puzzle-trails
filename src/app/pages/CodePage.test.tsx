import { SnackbarProvider } from "notistack";
import * as React from "react";

import render from "../../common/testing/render";

import CodePage from "./CodePage";

it("renders correctly", () => {
  const { container } = render(<CodePage codeId="foobar" />);
  expect(container.firstChild).toMatchSnapshot();
});
