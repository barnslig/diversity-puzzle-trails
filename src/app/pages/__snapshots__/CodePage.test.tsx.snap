// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorInherit mui-fixed MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <div
        class="MuiBox-root MuiBox-root-3"
      >
        <a
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeStart"
          href="/"
          tabindex="0"
          title="QR-Code nicht ausführen"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </a>
      </div>
      <h1
        class="MuiTypography-root makeStyles-appBarTitle-1 MuiTypography-h6"
      >
        QR-Code Aktionen
      </h1>
    </div>
  </header>
  <main
    class="MuiBox-root MuiBox-root-4"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
      >
        Dieser QR-Code führt folgende Aktionen aus:
      </p>
      <ul
        class="MuiList-root MuiList-padding"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <div
            class="MuiListItemIcon-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-pulse"
              style="width: 24px; height: 24px;"
            />
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-multiline"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                style="width: 24px; height: 24px;"
              />
            </span>
            <p
              class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                style="width: 140px; height: 20px;"
              />
            </p>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <div
            class="MuiListItemIcon-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-pulse"
              style="width: 24px; height: 24px;"
            />
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-multiline"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                style="width: 24px; height: 24px;"
              />
            </span>
            <p
              class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                style="width: 140px; height: 20px;"
              />
            </p>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <div
            class="MuiListItemIcon-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-pulse"
              style="width: 24px; height: 24px;"
            />
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-multiline"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                style="width: 24px; height: 24px;"
              />
            </span>
            <p
              class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
                style="width: 140px; height: 20px;"
              />
            </p>
          </div>
        </li>
      </ul>
      <div
        class="MuiBox-root MuiBox-root-5"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge Mui-disabled MuiButton-fullWidth Mui-disabled"
          disabled=""
          tabindex="-1"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Ausführen
          </span>
        </button>
      </div>
    </div>
  </main>
</div>
`;
